lms/djangoapps/experiments/views.py

```diff
@@ -84,3 +90,14 @@ class ExperimentKeyValueViewSet(viewsets.ModelViewSet):
     permission_classes = (IsStaffOrReadOnly,)
     queryset = ExperimentKeyValue.objects.all()
     serializer_class = serializers.ExperimentKeyValueSerializer
+
+
+class UserMetaDataView(APIView):
+
+    def get(self, request, course_id=None, username=None):
```

Comment:
Is there a reason why this is one endpoint? It looks to me like it's doing a best-effort attempt to get user data, then best-effort attempt to get course data. Why not have one endpoint to get user data, then another endpoint to get course data? Having them separate would allow for more reusability/discoverability (it's not clear that course data is a subset of user metadata, for instance), especially if this endpoint is not relying on having both pieces of input data.

Reply:
Yes this is by necessity; a bunch of stuff in the user-metadata relates to that user's enrollment in that course (like upgrade deadline, etc). That's why setting the context via get_experiment_user_metadata_context takes both a course and a user object together, and sets the user-metadata accordingly. Let me know if this doesn't make sense.

Reply:
If both the username and course_id are required, then should the endpoint return an error if one/both are missing?

Reply:
Yes probably! Good call- I'll update accordingly

Reply:
Added! It should only happen if we get something wrong while testing with Optimizely, but it's more complete and this way there'll be no mystery :)